### PR TITLE
Add html option to keep the forward slash in self closing tags

### PR DESCRIPTION
**Pull request by luchev**
Some older HTML clients/rendering engines cannot properly parse tags, which should be self-closing, but don't have the forward slash before the >.
I want to add an option so users of the library can choose to keep the forward slashes in such situations. This change should cause no degradation in functionality.

### CODE REVIEW

**Comment by tdewolff** [comment]
Thanks for the PR! In general, the inflexibility of requiring self-closing tags to have the end slash is very typical of an XML parser and not an HTML parser. Even the oldest HTML has allowed omitting the ending slash, since self-closing tags are defined in the specification. So it seems that your clients / rendering engines are not HTML but XML, and I'd urge you to use the XML minifiers since there are many small differences between HTML and XML (and you'll probably hit another roadblock after fixing this issue). The idea is not that the HTML minifier becomes an XHTML minifier as well.

However having said that, in your e-mail you described problems with `<svg>` tags, which is a different matter all together! Embedded SVG (and MathML) within HTML is actually XML, which starts at the begin tag and ends at the final tag, such that anything within `<svg ...> ... </svg>` is XML. However, I'm not sure how the SVG tags apply to your problem of self-closing tags, since `<svg .../>` is quite useless (but should contain the self-closing slash nonetheless).

**Comment by luchev** [comment]
It's not really a self-closing SVG that I have, it's the contents that are causing the problem. I should've gone into more detail instead of misleading you with the SVG. The root of the issue is that the HTML I have is using VML. This VML has such self-closing tags, that don't work properly in older email clients without the slash, e.g. `<v:fill type="frame" .... />`.

In the end, as you suggested, my issue is that I have some form of XML embedded in the HTML.

**Comment by tdewolff** [comment]
Yes, it looks like VML is based on XML. Reading https://www.w3.org/TR/NOTE-VML#h2:introduction it seems that VML has nothing to do with HTML (it is based on XML not on SGML). I don't think this minifier will support minifying VML files specifically, but the XML minifier should work sufficiently in this case. A quick search for VML in HTML4 or HTML5 spec didn't return anything, but if you find anything please let me know!

In short: you're trying to minify an XML document (which may look like an HTML document) with an HTML minifier. In any case, can you attach a small example of such a HTML/VML file that you're minifying, so I can get a better idea of the problem?

**Comment by luchev** [comment]
I tried using the XML minifier for the whole document but it doesn't work for my use case. I am really minifying an HTML document with its CSS, except it has parts that are XML.

Here's an example VML as a part of a bigger HTML document. I've removed all the html, body, etc. tags and left a relevant and small example:
```
<div>
  <!--[if gte mso 9]>
  <v:rect xmlns:v="urn:schemas-microsoft-com:vml" fill="true">
  <v:fill type="frame" origin="0, 1" position="0,1" src="myimage.png" />
  <![endif]-->
</div>
```

The bigger issue is XML embedded in HTML under some form, not exactly the self-closing tags.
I understand that making a flag for self-closing tags is a very specific change and it doesn't address the bigger issue in its entirety. But I think if we looked into minifying XML, which is embedded into HTML, it might become a very big effort.

**Comment by tdewolff** [comment]
I'm a bit at blank here. I've been looking all over the HTML4 and HTML5 specifications, but there doesn't seem to be any mention of VML. On the other hand, mentions

>  The first part of this document is an introduction, which gives an overview of the way VML is organized and how it interacts with both XML and HTML as defined by the HTML 4.0 Specification.

But this seems to be false, and later on they show an image of using VML (within XML) or using HTML (without VML) to draw graphics. It does seem to talk about embedding HTML within VML, but not the other way around. In fact, it seems that Outlook doesn't really support HTML5 (and 4?), so I think it really is parsing XHTML (which is XML). You probably have to add `xmlns="http://www.w3.org/1999/xhtml` to the `<html>` tag, which tells us this is really XML!

In any case, you probably want to send XHTML with VML to Outlook clients, but HTML4/5 to other clients. This divide may have more complications than just some VML ;-). For practical purposes though, supposing that this is a common problem, we could add an exception for VML so that it will be parsed as XML (just the self-closing tags part will introduce other problems later on, XML and HTML have many more differences!). Cna we safely say that any VML element starts with `v:`? Can you show me a more complete version of a document with VML?

**Comment by luchev** [comment]
I went through the specification briefly as well. The only reference of what we're using I can find on [wikipedia](https://en.wikipedia.org/wiki/Vector_Markup_Language#:~:text=instead%20of%20VML.-,Microsoft%20Outlook%20HTML%20email%20rendering,-%5Bedit%5D). It's really strange that isn't mentioned in the official specification.

I agree adding the logic for self-closing tags is very hacky. I didn't quite understand the scope of the issue when I started investigating it.

I think that VML elements start with `v:`, but I can't be 100% certain. I'm knowledgeable enough about VML. I don't really like the idea of adding exceptions for such outdated and deprecated technologies. This isn't used anywhere except in outlook emails from something like 2010 : )

I can provide a bigger document, but there's literally nothing more to it. The rest is a typical HTML. We don't use anything specific. The document starts with `<!doctype html><html>`. The code around the `div` is just more divs and tables. There's no more VML in the document.

Let's close this pull request. I'll find another solution to the problem on our side as it's something very specific.
Thank you for all the time you spent on this.
